### PR TITLE
[Winforms] Form fixes

### PR DESCRIPTION
**Pull request by PreferLinux**
- In SelectActiveControl(), use SendControlFocus() rather than Select() because the former simply gives it focus, the latter will set ActiveControl which is at best unnecessary, and I think can cause problems (I'm sure I've had some, but can't recall the details).
- In CreateHandle(), ensure the ClientSize is correct if it is set. If not, update it to make it right. Suspend layout while changing it so that child controls with anchors other than Top Left are not moved. This is important because forms are normally sized with ClientSize, and on X11 when they are initially created they are sized based on estimated borders (border sizes from Theme, caption height from XplatUIX11). Once they're created, the actual client size can be calculated, and from there the correct form size.
- When pressing Enter on a form, if a button is selected, ensure it is both visible and enabled before doing PerformClick. Shouldn't be selected in the first place (I've submitted a fix for that in #18426), but...

### CODE REVIEW

**Comment by akoeplinger** [comment]
@monojenkins build

**Comment by akoeplinger** [comment]
@PreferLinux this seems to cause some unit test failures:

```
- FormTest.AutoScaleBounds
  A0
  Expected: {Width=213, Height=121}
  But was:  {Width=368, Height=209}

-  FormTest.bug_82358
  FixedToolWindow-E3
  String lengths are both 23. Strings differ at index 7.
  Expected: "{Width=198, Height=195}"
  But was:  "{Width=298, Height=295}"
  ------------------^
```

**Comment by PreferLinux** [comment]
OK, I know why those are happening. In both cases ClientSize has been set, and then the actual bounds have been changed (either by AutoScale calling SetBounds with BoundsSpecified.All, or by directly setting Size). After that, this code sees `is_clientsize_set == true`, so sets ClientSize to `clientsize_set`, undoing the previous change.

Of note, setting Form.Menu has exactly the same bug, although it checks `clientsize_set != Size.Empty`.

In my opinion the problem is that `is_clientsize_set` remains true even after other things change the desired bounds. The difficulty is determining where the change should clear it and where it shouldn't.

**Comment by PreferLinux** [comment]
Well, I've definitely seen problems that that fixed, but it definitely was not the correct fix. And my opinion of what happens was also definitely wrong, because stuff does size and position correctly with varying border and caption sizes on Linux / X11 – even huge ones.

The problems I have seen happen with only some BorderStyles, and only when ControlBox is false. And the form can end out significantly smaller than it should be. So I'll go digging to try and figure it out.

**Comment by PreferLinux** [comment]
So subsequent to the above, I've investigated further. It actually happened with any FormBorderStyle other than None. And only some sequences of setting the form Text, ControlBox, FormBorderStyle and ClientSize properties caused problems and others didn't. I believe I've tracked it down to Form.CreateParams lines 1486-1487, specifically
```
if (!ControlBox && Text == string.Empty) {
	cp.WindowStyle &= ~WindowStyles.WS_DLGFRAME;
}
```
Commenting out that seems to prevent the problem, but I'm wondering just what that is meant to do.

**Comment by akoeplinger** [comment]
@filipnavara do you have any idea about this?

**Comment by PreferLinux** [comment]
I've still got no idea why that is being done, but I have investigated further and come to a couple of conclusions.

First, when that clears `WS_DLGFRAME` like that, it also removes `WS_CAPTION` since `WS_CAPTION == WS_BORDER | WS_DLGFRAME`. That causes incorrect border measurements in Hwnd.GetBorders, and possibly has other significant effects too. I don't see why that would ever be wanted. However, I've found the commit the change was made in: https://github.com/mono/mono/commit/a1d276ed06de5e07c3adaf15ca98350774d5b807 I can't find bug 79386...

Secondly, it sure seems like this is the source. Setting FormBorderStyle isn't even necessary – it'll happen with this test code (the button is just to show the correct bottom-right corner and provide a convenient close):
```C#
using System.Windows.Forms;
class testform : Form {
	static void Main() {
		Application.Run(new testform());
	}
	
	testform() {
		ControlBox = false;
		ClientSize = new System.Drawing.Size(270, 170);
		Text = "Test Form";
		
		var btn = new Button();
		btn.SetBounds(200, 70, 70, 100);
		// btn.Right == ClientSize.Width and btn.Bottom == ClientSize.Height
		Controls.Add(btn);
		btn.Click += (s, e) => Close();
	}
}
```
Basically, setting ClientSize will get CreateParams, and use that to find the correct window size. Unfortunately, because this does that after setting ControlBox false but before setting Text, the CreateParams returned doesn't have WS_CAPTION set. So it calculates an incorrect window size.

**Comment by filipnavara** [comment]
> First, when that clears `WS_DLGFRAME` like that, it also removes `WS_CAPTION` since `WS_CAPTION == WS_BORDER | WS_DLGFRAME`. That causes incorrect border measurements in Hwnd.GetBorders, and possibly has other significant effects too. I don't see why that would ever be wanted. However, I've found the commit the change was made in: [a1d276e](https://github.com/mono/mono/commit/a1d276ed06de5e07c3adaf15ca98350774d5b807) I can't find bug 79386...

Yep, that looks wrong. The bug number is probably the Novell bugzilla which is not accessible anymore.

For the record, here is the dotnet/winforms code for the similar logic:

https://github.com/dotnet/winforms/blob/4a69ae682f19957d0df8179885104631cf4a323b/src/System.Windows.Forms/src/System/Windows/Forms/Form.cs#L3454-L3468

`!string.IsNullOrEmpty(Text) || ControlBox` should imply `WS_CAPTION`. It doesn't make any sense to manipulate `WS_DLGFRAME` individually.

**Comment by PreferLinux** [comment]
> > First, when that clears `WS_DLGFRAME` like that, it also removes `WS_CAPTION` since `WS_CAPTION == WS_BORDER | WS_DLGFRAME`. That causes incorrect border measurements in Hwnd.GetBorders, and possibly has other significant effects too. I don't see why that would ever be wanted. However, I've found the commit the change was made in: [a1d276e](https://github.com/mono/mono/commit/a1d276ed06de5e07c3adaf15ca98350774d5b807) I can't find bug 79386...
> 
> Yep, that looks wrong. The bug number is probably the Novell bugzilla which is not accessible anymore.
> 
> For the record, here is the dotnet/winforms code for the similar logic:
> 
> https://github.com/dotnet/winforms/blob/4a69ae682f19957d0df8179885104631cf4a323b/src/System.Windows.Forms/src/System/Windows/Forms/Form.cs#L3454-L3468
> 
> `!string.IsNullOrEmpty(Text) || ControlBox` should imply `WS_CAPTION`. It doesn't make any sense to manipulate `WS_DLGFRAME` individually.

After trying the no Text or ControlBox case on Windows, I think I know why that is there – on Windows that results in a window with no title / caption, but with a border (unlike FormBorderStyles.None). Mono's `CreateParams` puts `WS_CAPTION` on when dealing with FormBorderStyle, so that case is handled by removing `WS_DLGFRAME` instead. So yeah, looks weird, but I think it is correct.

> > Basically, setting ClientSize will get CreateParams, and use that to find the correct window size. Unfortunately, because this does that after setting ControlBox false but before setting Text, the CreateParams returned doesn't have WS_CAPTION set. So it calculates an incorrect window size.
> 
> Anything that affects the window decorations should also cause the recalculation of the bounds. I'd expect the `Text` and `ControlBox` setters to do that. The dotnet/winforms code linked above does that by calling `UpdateFormStyles()` method for all the problematic cases (eg. switching from non-empty text to empty one, modifying ControlBox, etc.).
> 
> The method looks like this:
> 
> ```
>         internal void UpdateFormStyles()
>         {
>             Size previousClientSize = ClientSize;
>             base.UpdateStyles();
>             if (!ClientSize.Equals(previousClientSize))
>             {
>                 ClientSize = previousClientSize;
>             }
>         }
> ```
> 
> (`UpdateStyles` already exists in Mono code and is called from some of the places)

`ControlBox` does call `UpdateStyles`, but the first thing `UpdateStyles` does is check `IsHandleCreated`, and if false (which it is – we're in the form constructor), returns. `Text` calls `UpdateWindowText` which, once again, doesn't do anything if the handle isn't created. `FormBorderStyle` also calls `UpdateStyles`, and then matches up the client sizes itself, but only if either the handle is created or the client size has been set.

`ClientSize` calculates the window size with one `CreateParams`, and then CreateHandle creates the window using a different `CreateParams` – which on X11 uses overall size, and calculates the X window size by taking off borders according to `CreateParams`. But as you say, the real problem is that bounds haven't been recalculated correctly.

**Comment by PreferLinux** [comment]
> > Basically, setting ClientSize will get CreateParams, and use that to find the correct window size. Unfortunately, because this does that after setting ControlBox false but before setting Text, the CreateParams returned doesn't have WS_CAPTION set. So it calculates an incorrect window size.
> 
> Anything that affects the window decorations should also cause the recalculation of the bounds. I'd expect the `Text` and `ControlBox` setters to do that. The dotnet/winforms code linked above does that by calling `UpdateFormStyles()` method for all the problematic cases (eg. switching from non-empty text to empty one, modifying ControlBox, etc.).
> 
> The method looks like this:
> 
> ```
>         internal void UpdateFormStyles()
>         {
>             Size previousClientSize = ClientSize;
>             base.UpdateStyles();
>             if (!ClientSize.Equals(previousClientSize))
>             {
>                 ClientSize = previousClientSize;
>             }
>         }
> ```
> 
> (`UpdateStyles` already exists in Mono code and is called from some of the places)

OK, I've done something like that. I'll commit and push that change tomorrow, and see how it goes.

**Comment by PreferLinux** [comment]
Right, I hope this works properly. I can't guarantee that I've got everything that could change window decoration sizes, and I'm wondering about MDI stuff in particular.

**Comment by filipnavara** [comment]
The ClientSize part looks ok to me now, the rest I didn't check.

**Comment by sancheolz** [comment]
@PreferLinux, maybe yours PR fixed issue #10642. Thank you for contribution.
